File: othello_python/othello/colorprint.py
"""
Colorprint
Color print helper
Akseli Lukkarila
2019-2023
"""
import sys

import colorama

colorama.init()


class Color:
    """Interface to Colorama colors."""

    blue = colorama.Fore.BLUE
    cyan = colorama.Fore.CYAN
    green = colorama.Fore.GREEN
    magenta = colorama.Fore.MAGENTA
    red = colorama.Fore.RED
    white = colorama.Fore.WHITE
    yellow = colorama.Fore.YELLOW


def get_color(text: str, color=Color.white, bold=False) -> str:
    """Format string with colour using Colorama."""
    return f"{colorama.Style.BRIGHT if bold else ''}{color}{text}{colorama.Style.RESET_ALL}"


def print_bold(text: str, color=Color.white, **kwargs):
    """Print bold text."""
    print(get_color(text, color, True), **kwargs)


def print_color(text: str, color=Color.white, bold=False, **kwargs):
    """
    Print text with colour using Colorama.

    You can pass optional extra arguments for the Python default print function if wanted:

    - sep='separator': Specify how to separate the objects, if there is more than one.
        Default is ' '
    - end='end': Specify what to print at the end. Default is '\\n' (line feed)
    - file: An object with a write method. Default is sys.stdout
    - flush: A Boolean, specifying if the output is flushed (True) or buffered (False).
        Default is False
    """
    print(get_color(text, color, bold), **kwargs)


def print_error_and_exit(message: str, exit_code=1):
    """Print error message and exit with given exit code."""
    print_error(message)
    sys.exit(exit_code)


def print_error(message: str, bold=False, **kwargs):
    """Print error message with red colour."""
    indent, text = _split_leading_whitespace(message)
    print_color(f"{indent}ERROR: {text}", Color.red, bold, **kwargs)


def print_warn(message: str, bold=False, **kwargs):
    """Print warning message with yellow colour."""
    indent, text = _split_leading_whitespace(message)
    print_color(f"{indent}WARNING: {text}", Color.yellow, bold, **kwargs)


def print_green(text: str, bold=False, **kwargs):
    print_color(text, Color.green, bold, **kwargs)


def print_yellow(text: str, bold=False, **kwargs):
    print_color(text, Color.yellow, bold, **kwargs)


def print_red(text: str, bold=False, **kwargs):
    print_color(text, Color.red, bold, **kwargs)


def print_blue(text: str, bold=False, **kwargs):
    print_color(text, Color.red, bold, **kwargs)


def print_magenta(text: str, bold=False, **kwargs):
    print_color(text, Color.magenta, bold, **kwargs)


def print_cyan(text: str, bold=False, **kwargs):
    print_color(text, Color.cyan, bold, **kwargs)


def _split_leading_whitespace(message: str) -> (str, str):
    """Split a string to the leading whitespace and rest of the string."""
    indent_size = next((i for i, c in enumerate(message) if not c.isspace()), len(message))
    indent = message[:indent_size]
    text = message[indent_size:]
    return indent, text

File: othello_kotlin/app/src/main/kotlin/othello_kotlin/ColorPrint.kt
package othello_kotlin

import java.awt.Color

/** Format string with colour.*/
fun getColor(text: String, color: Color): String {
    val ansiColor = when (color) {
        Color.CYAN -> "\u001B[36m"
        Color.GREEN -> "\u001B[32m"
        Color.MAGENTA -> "\u001B[35m"
        Color.RED -> "\u001B[31m"
        Color.WHITE -> "\u001B[37m"
        Color.YELLOW -> "\u001B[33m"
        // Default to white
        else -> "\u001B[37m"
    }
    // \u001B[0m resets the colour
    return "$ansiColor$text\u001B[0m"
}

/** Print text with specified colour.*/
fun printColor(text: String, color: Color) {
    println(getColor(text, color))
}

/** Print bold text.*/
fun printBold(text: String) {
    println("\u001B[1m${text}\u001B[0m")
}

/** Print error message with red colour.*/
fun printError(message: String) {
    val (indent, text) = splitLeadingWhitespace(message)
    println("$indent${getColor("Error: $text", Color.RED)}")
}

/** Print warning message with yellow colour.*/
fun printWarn(message: String) {
    val (indent, text) = splitLeadingWhitespace(message)
    println("$indent${getColor("Warning: $text", Color.YELLOW)}")
}

/** Split a string into the leading whitespace and the rest of the string.*/
fun splitLeadingWhitespace(message: String): Pair<String, String> {
    val indentSize = message.takeWhile { it.isWhitespace() }.length
    return message.substring(0, indentSize) to message.substring(indentSize)
}
